cmd/podman/compose.go

```diff
@@ -0,0 +1,264 @@
+	RunE:              composeMain,
+	ValidArgsFunction: completion.AutocompleteNone,
+	Example: `podman compose -f nginx.yaml up --detach
+  podman --log-level=debug compose -f many-images.yaml pull`,
+	DisableFlagParsing: true,
+	Annotations:        map[string]string{registry.ParentNSRequired: ""}, // don't join user NS for SSH to work correctly
+}
+
+func init() {
+	// NOTE: we need to fully disable flag parsing and manually parse the
+	// flags in composeMain. cobra's FParseErrWhitelist will strip off
+	// unknown flags _before_ the first argument.  So `--unknown argument`
+	// will show as `argument`.
+
+	registry.Commands = append(registry.Commands, registry.CliCommand{Command: composeCommand})
+}
+
+// composeProvider provides the name of or absolute path to the compose
+// provider (i.e., the external binary such as docker-compose).
+func composeProvider() (string, error) {
+	if value, ok := os.LookupEnv("PODMAN_COMPOSE_PROVIDER"); ok {
+		return value, nil
+	}
+
+	candidates := registry.PodmanConfig().ContainersConfDefaultsRO.Engine.ComposeProviders
+	if len(candidates) == 0 {
+		return "", errors.New("no compose provider specified, please refer to `man podman-compose` for details")
+	}
+
+	for _, candidate := range candidates {
+		path, err := exec.LookPath(candidate)
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				continue
+			}
+			return "", err
+		}
+		// First specified provider "candidate" wins.
+		logrus.Debugf("Found compose provider %q", path)
+		return path, nil
+	}
+
+	return "", errors.New("no configured compose provider found on system, please refer to `man podman-compose` for details")
+}
+
+// composeEnv returns the compose-specific environment variables.
+func composeEnv() ([]string, error) {
+	hostValue, err := func() (string, error) {
+		if value, ok := os.LookupEnv("DOCKER_HOST"); ok {
+			return value, nil
+		}
+
+		// For local clients (Linux/FreeBSD), use the default API
+		// address.
+		if !registry.IsRemote() {
+			return registry.DefaultAPIAddress(), nil
+		}
+
+		cfg, err := config.ReadCustomConfig()
+		if err != nil {
+			return "", err
+		}
+
+		// NOTE: podman --connection=foo and --url=... are injected
+		// into the default connection below in `root.go`.
+		defaultConnection := cfg.Engine.ActiveService
+		if defaultConnection == "" {
+			switch runtime.GOOS {
+			// If no default connection is set on Linux or FreeBSD,
+			// we just use the local socket by default - just as
+			// the remote client does.
+			case "linux", "freebsd":
+				return registry.DefaultAPIAddress(), nil
+			// If there is no default connection on Windows or Mac
+			// OS, we can safely assume that something went wrong.
+			// A `podman machine init` will set the connection.
+			default:
+				return "", fmt.Errorf("cannot connect to a socket or via SSH: no default connection found")
```

Comment:
Since we're already switching on GOOS here, it feels like we can safely recommend `podman machine init` to get a VM running in this error